experimental/pattern-matching.md

```diff
@@ -0,0 +1,126 @@
+# [Pattern Matching](proposal-pattern-matching)
+
+See also [AST examples](ast-examples) for this proposal.
+
+## MatchExpression
+
+```js
+interface MatchExpression <: Expression {
+    type: "MatchExpression";
+    discriminant: Expression;
+    id: Pattern | null;
+    clauses: [ MatchClause ]
+}
+```
+
+## WhenClause
+```js
+interface MatchClause <: Node {
+    type: "MatchClause";
+    test: MatchPattern | null;
+    guard: Expression | null;
+    id: Pattern | null;
+    consequent: BlockStatement
+}
+```
+
+If `test` is `null`, `id` must be `null`.
+
+## MatchPattern
+```js
+interface MatchPattern <: Node {}
+```
+
+While a Match Pattern shares similar structures with a Pattern, it is different from a Pattern:
+
+- Literals or negative numbers are match patterns, but they are not patterns
+- Any expression can be converted to an expression match pattern
+- Multiple match patterns can be joined with `and`, `or` and `with` operators, but patterns can not
+- Unlike array and object patterns, assignment patterns have no corresponding syntax in match patterns
+
+## ArrayMatchPattern
+```js
+interface ArrayMatchPattern <: MatchPattern {
+    type: "ArrayMatchPattern";
+    elements: [ MatchPattern ];
+}
+```
+
+## ObjectMatchPattern
+```js
+interface AssignmentMatchProperty <: Property {
+    type: "Property"; // inherited
+    value: MatchPattern;
+    kind: "init";
+    method: false;
+}
+
+interface ObjectMatchPattern <: MatchPattern {
+    type: "ObjectMatchPattern";
+    properties: [ AssignmentMatchProperty ];
```

Comment:
```suggestion
    properties: [ AssignmentMatchProperty | Property ];
```

Reply:
Can you elaborate? `AssignmentMatchProperty` inherits `Property` but its value is limited to a `MatchPattern`. A general object property can not fit here, e.g.

```js
match (x) {
  when ({ a: () => {} }) {}
}
```
is invalid because `() => {}` is not a match pattern.

Reply:
Sure, sorry. If `ObjectMatchPattern` only accepts `AssignmentMatchProperty` as its properties, that means that even
`{ status: 200 | 201 }` should have `"kind": "init"`. But `status` should not create a new binding, right?

Reply:
Yes. In ESTree a pattern element in an ObjectPattern always has `"kind": "init"`.  The ObjectMatchPattern is aligned with ObjectPattern. 

> `status` should not create a new binding

Nope. A match pattern creates bindings (if matched). See the code examples on the proposal:

```js
match (res) {
  if (isEmpty(res)) { … }
  when ({ data: [page] }) { … }
  when ({ data: [frontPage, ...pages] }) { … }
  else { … }
}
```

> The first when clause matches if data has exactly one element, and binds that element to page for the right-hand side.

Reply:
My doubts concern the examples provided above where `Property` is used instead of `AssignmentMatchProperty`.

Reply:
I am not sure I can follow. Can you elaborate why `properties` of an `ObjectMatchPattern` should accept general `Property`?

The `type` of `AssignmentMatchProperty` is `"Property"`. It aligns to how ESTree models an object pattern: `a: 1` in `({ a: 1 } = {})` is a `property`. See also https://github.com/estree/estree/blob/master/es2015.md#objectpattern

Reply:
It's clear now, I though that `AssignmentMatchProperty` would have been a new node🤦